cpp/src/groupby/sort/group_shift.cu

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/groupby/group_shift.hpp>
+#include <cudf/detail/groupby/sort_helper.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/scatter.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+
+namespace {
+
+constexpr size_type SAFE_GATHER_IDX = 0;
+
+/**
+ * @brief Functor to determine the location to set `fill_value` for groupby shift.
+ */
+template <bool ForwardShift, typename EdgeIterator>
+struct group_shift_fill_functor {
+  EdgeIterator group_edges_begin;
+  size_type offset;
+  size_type group_label, offset_to_edge;
+
+  group_shift_fill_functor(EdgeIterator group_edges_begin, size_type offset)
+    : group_edges_begin(group_edges_begin), offset(offset)
+  {
+  }
+
+  __device__ size_type operator()(size_type i)
+  {
+    if (ForwardShift) {  // offset > 0
+      group_label    = i / offset;
+      offset_to_edge = i % offset;
+    } else {  // offset < 0
+      group_label    = -i / offset;
+      offset_to_edge = i % offset + offset + 1;
+    }
+    return *(group_edges_begin + group_label) + offset_to_edge;
+  }
+};
+
+}  // namespace
+
+/**
+ * @brief Implementation of groupby shift
+ *
+ * Groupby shift is based on sort groupby. The first step is a global shift for `sorted_values`.
+ * The second step is to set the proper locations to `fill_values`.
+ *
+ * @tparam EdgeIterator Iterator type to the group edge list
+ *
+ * @param sorted_values values to be sorted, grouped by keys
```

Comment:
values should not be sorted. fix (naming and documentation) at other places too.